### DISCUSSION TITLE
How do I create a function that takes in a variable and returns a new StaticQueryDocument in Gatsby & GraphQL

### DISCUSSION THREAD

**Discussion by do-kevin** [gatsbyjs/gatsby#34911] [Help] [answered]
Hey there,

I'm in a pickle figuring out how to query for my images. I am using [ ].map() to render React elements that have `<StaticQuery />` in them. I want to give `StaticQuery` a new query, as the loop progresses, that will fetch my required image. 

Example array:
```
[
  {
    title: "Lorem ipsum",
    description: "What am I doing",
    imageFileName: "brand.png"
  },
 {
    title: "Lorem ipsum 2",
    description: "This is confusing",
    imageFileName: "brand2.png"
 }
]
```

Query I want to use and modify:
```
graphql`query {
    image: file(relativePath: { eq: "images/brand.png" }) {
      childImageSharp {
        fluid(maxWidth: 200, maxHeight: 200) {
          ...GatsbyImageSharpFluid
        }
      }
  },
}`
 ```
 
I want to generate a new query by editing `relativePath`'s `eq` by giving it a different image path and file name. I've tried to use template literals, but Gatsby and TypeScript gave me errors. I've looked at some info about gatsby-node and createResolvers, but I am quite confused on how to create a resolver if that's the right approach I'm supposed to be doing.

How do I create this? Could you give me some example simple code I could look at?
 
gatsby ^4.3.0
typescript ^4.5.4

**Comment by psykzz** [comment, marked as answer]
As you may have already experienced StaticQuery doesn't take arguments.
It may help to understand what is happening behind the scenes to explain why this isn't possible.

When you do a query (for example)
```graphql
query {
    images: files {
      childImageSharp {
        fluid(maxWidth: 200, maxHeight: 200) {
          ...GatsbyImageSharpFluid
        }
     }
  }
}
```

What is happening here is
1. Gatsby finds all the queries
2. Using the query runner gatsby saves the results as JSON to disk.
3. Gatsby replaces all the useStaticQuery, or Provider with a more complex version `return await fetch('./query-response-123.json).json()`.

The reason this is important is because the data within the response is already set in stone.

What you may find easier is to just query down to the minimum dataset possible and then just do the filtering in your components.

For example if you add `relativePath` to your query above, you could do something like


```js
const getImageByPath(path: string) {
  const data = useStaticQuery(graphql`
    query {
        images: files {
          relativePath
          childImageSharp {
            fluid(maxWidth: 200, maxHeight: 200) {
              ...GatsbyImageSharpFluid
            }
         }
      }
    }
  `)
  return data.images.find(image => image.relativePath === path);
}
```

Depending on how many images you have this _could_ be a large file, but if you have only say a few hundred images, this would work for most cases.


_Note i'm hoping this can be applied generically to your problem. Obviously things are nuanced and I don't have all the context._

**Reply by do-kevin** [reply]
Thanks for explaining the process. Yeah, I ended up  having to fetch all the files in a folder with the query and used the returning data to compose an object using `_.keyBy()` that'll be matched up against each item in my loop.

**Reply by LekoArts** [reply]
I just want to point out here that this is a big anti-pattern and you'll be in great trouble if you use this long-term. Make your components "dumb" and pass in the data that you query on **pages** or use something like `<StaticImage />`.

**Reply by do-kevin** [reply]
> I just want to point out here that this is a big anti-pattern and you'll be in great trouble if you use this long-term. Make your components "dumb" and pass in the data that you query on **pages** or use something like `<StaticImage />`.

So this is the code I repeatedly implemented so far. A bunch of StaticQueries in my project's page like this one for example. Is this okay or is it an anti-pattern, and/or does not scale well?

```
import { graphql, StaticQuery } from "gatsby";
import BackgroundImage from "gatsby-background-image";
import { keyBy } from "lodash";

const exampleArray = [
  {
    title: "Lorem ipsum",
    description: "What am I doing",
    imageFileName: "brand.png"
  },
 {
    title: "Lorem ipsum 2",
    description: "This is confusing",
    imageFileName: "brand2.png"
 },
 ....etc... like 6 more items
]

const ScreenComponent = () => {
   let myMap = {};
   return (
      <div>
         {exampleArray.map(info => {
             return (
                 <StaticQuery
                     query={graphql`
                            query {
                                   allFile(filter: { relativeDirectory: { eq: "images" } }) {
                                      nodes {
                                          childImageSharp {
                                              fluid(fit: COVER) {
                                                 ...GatsbyImageSharpFluid
                                                    originalName
                                    }
                                 }
                              }
                           }
                        }
                     `}
                     render={(data) => {
                        if (Object.keys(myMap).length !== exampleArray.length) {
                           myMap = keyBy(data.allFiles.nodes, "childImageSharp.fluid.originalName");
                        }
                        
                        const imageData = myMap[info.imageFileName].childImageSharp.fluid;
                        
                        return (
                           <BackgroundImage fluid={imageData}>
                               <p>{info.title}</p>
                               <p>{info.description}</p>
                           </BackgroundImage>
                        );
                     } }
                 />
             );
          }) ;
         }
      </div>
   );
}
```
